sympy/functions/elementary/hyperbolic.py

```diff
@@ -681,13 +681,17 @@ def _eval_rewrite_as_coth(self, arg, **kwargs):
         return 1/coth(arg)
 
     def _eval_as_leading_term(self, x, logx=None, cdir=0):
-        from sympy.series.order import Order
         arg = self.args[0].as_leading_term(x)
+        arg0 = arg.subs(x, 0)
 
-        if x in arg.free_symbols and Order(1, x).contains(arg):
+        if arg0 is S.NaN:
+            arg0 = arg.limit(x, 0, dir='-' if cdir.is_negative else '+')
```

Comment:
Can cdir be - 1?

Reply:
I took inspiration for this from the `sinh` function but I don't think we have a test yet where we encounter a case where `arg0` is `S.NaN`, so not too sure on this but if you have anything in mind let me know , i'll check it out ! I'll try to think when could we encounter such case till then !

Reply:
> the suggestion you gave me

Do you mean the suggestion to use an improved version of `log`?

Reply:
I'm actually a bit confused about this comment now , can we do all log operations at all places( not that I know any restrictions) but I came across some results which make me oppose this simplifcation here !
For eg
```
>>> asech(x).rewrite(log)
log(sqrt(-1 + 1/x)*sqrt(1 + 1/x) + 1/x)

>>> expr = log(x*sqrt(-1 + 1/x)*sqrt(1 + 1/x) + 1) # basically simplying log(x) + asech(x)

>>> limit(asech(x).rewrite(log) + log(x), x, 0,'-')
log(2) + 2*I*pi 

>>> limit(expr, x, 0, '-')
log(2) # This answer opposes the one from wolfram and few other CAS(sage for example)
```

Reply:
I'll paste a set of results which I feel are the correct ones (after confirming with wolfram and other CAS)
Ignore the rewrite required in these, all results are correct but in sympy we could only get the correct results as of now on master using the rewrite ! But other than that wolfram agrees to all these answers !
```
>>> limit(log(x) + asech(x).rewrite(log), x,0, '-')
log(2) + 2*I*pi
>>> limit(log(x) + asech(x).rewrite(log), x,0, '+')
log(2)
>>> limit(-log(x) + asech(x).rewrite(log), x,0, '-')
oo
>>> limit(-log(x) + asech(x).rewrite(log), x,0, '+')
oo
>>>
>>>
>>> limit(log(x) + acsch(x).rewrite(log), x,0, '-')
-oo
>>> limit(log(x) + acsch(x).rewrite(log), x,0, '+')
log(2)
>>> limit(-log(x) + acsch(x).rewrite(log), x,0, '-')
-log(2) - I*pi
>>> limit(-log(x) + acsch(x).rewrite(log), x,0, '+')
oo
```

The best part is my first commit was giving all answers perfectly and after trying to consider the suggestion you pointed above now this is what we get
```
>>> limit(log(x) + asech(x), x, 0, '-')
log(2)
>>> limit(log(x) + asech(x), x, 0, '+')
log(2)
>>> limit(-log(x) + asech(x), x, 0, '-')
oo
>>> limit(-log(x) + asech(x), x, 0, '+')
oo
>>> limit(log(x) + acsch(x), x, 0, '-')
log(2)
>>> limit(log(x) + acsch(x), x, 0, '+')
log(2)
>>> limit(-log(x) + acsch(x), x, 0, '-')
-log(2) - I*pi
>>> limit(-log(x) + acsch(x), x, 0, '+')
oo

```
Some tests I had added in the first commit, which agree with all what wolfram says and the rewrite ones !
```
    assert limit(log(x)+asech(x),x,0, dir = '+') == log(2)
    assert limit(log(x)+asech(x),x,0, dir = '-') == log(2) + 2*I*pi
    assert limit(log(x)+acsch(x),x,0, dir = '+') == log(2)
    assert limit(log(x)+acsch(x),x,0, dir = '-') == -oo
```

Reply:
>It seems that this should be something like return log(2) - log(c) - e*log(x) when arg is c*x**e. If log(2) is not included, then the leading term of log(x) + asech(x) will not be correct. log(c*x**e) should be written in the form log(c) + e*log(x) so that the log(x) terms coming from different parts of a sum would combine without explicit simplifications.

I feel these should be the correct results 
```

>>> asech(x).as_leading_term(x)
log(2/x)
>>> asech(2*x).as_leading_term(x)
log(1/x)
>>> (2*asech(x)).as_leading_term(x)
2*log(2/x)
>>> ( log(x) + asech(x)).as_leading_term(x)
log(2/x) + log(x)  # If we cancel the x here , the essence of the imaginary part of the limit is being lost for some reason
```
Returning this will give us correct answer for limits !

Reply:
Once I have some clarity on this , I plan to do the following !
The `acsch(x)` implementation here doesn't trouble us though and works perfectly fine as confirmed with other CAS , cause the branch cuts are along the imaginary axis ! , hence once we've address the above pointed out thoughts 
i'll maybe create a new PR for the `acsch` one and we could continue our discussion here for how to handle the branch cuts for `asech` along real axis !

Reference from wolfram 
![image](https://user-images.githubusercontent.com/87052487/153228455-60481d3c-4247-437a-9512-be2e46ede41d.png)
![image](https://user-images.githubusercontent.com/87052487/153228570-19c337ec-5929-404e-a397-ac5b0d788710.png)

Reply:
It is possible that the two logarithms cannot be combined as in https://github.com/sympy/sympy/pull/22521#discussion_r753669192. That would explain the extra `2*pi`.

Reply:
Yeah I agree , I haven't given the log based approach your mentioning a fill fledged attempt ,will give it a shot in a day or two and see all areas where we could improve !

Reply:
The pole error, or at least some of them, should be removed and some log term returned instead.

Reply:
True true will try to approach those too ,I realized if we do some basic log operations as we had done before ( which lead to missing of the 2 pi term) in that case the as leading term works but in the purest form it won't! 
Will give it a shot by a day or two !

Reply:
Actually I realize that the pr is good to go by current standard(where I see that `_eval_as_leading_term`  which are already implemented for functions eg `asinh, acosh`  don't really work at branch cuts) but I wouldn't want that otherwise the essence of the discussion would be lost !
Check out this issue #22986
Here we have the following 
```
>>> limit(acosh(1 + 1/x)*sqrt(x), x, oo) # this comes from _eval_as_leading_term of acosh function
0                                       # acosh(x + 1).as_leading_term(x) is sqrt(2)*sqrt(x) but returns 0
>>> limit(acosh(1 + 1/x).rewrite(log)*sqrt(x), x, oo)
sqrt(2)
>>> gruntz(acosh(1 + 1/x)*sqrt(x), x, oo)
Traceback (most recent call last):
sympy.core.function.PoleError: Cannot expand acosh(_w + 1) around 0
>>> gruntz(acosh(1 + 1/x).rewrite(log)*sqrt(x), x, oo)
sqrt(2)
```
So basically this tells us that `acosh(x).series(x, 1, 6)` won't work ! Hence what changes are introduced in log need to cater to all functions also those which have been implemented . Obviously a basic fix for this issue is adding this block 

```
    def _eval_nseries(self, x, n, logx):
        x0 = self.args[0].limit(x, 0)
        if x0 == 1:
            return self._eval_rewrite_as_log(self.args[0])._eval_nseries(x, n, logx)
        else:
            return super()._eval_nseries(x, n, logx)
```
I had given a thought on the reported issue few days back and obviously this was solv

Reply:
**Will this solution work for the reported issue ?**
I ask this because 
**1)** The `_eval_nseries` implementation makes gruntz work ! So now we have
```
>>> gruntz(acosh(1 + 1/x)*sqrt(x), x, oo)
sqrt(2)
```
**2)** The `_eval_as_leading_term` improvement makes the leading term work and then as `leadterm` won't be independent of `x`  , we could return the working solution through gruntz
```
>>> acosh(1 + x).as_leading_term(x)
sqrt(2)*sqrt(x)
>>> 
>>> (acosh(1 + x)*sqrt(1/x)).as_leading_term(x)
sqrt(2)*sqrt(x)*sqrt(1/x)
>>> 
>>> (acosh(1 + x)*sqrt(1/x)).leadterm(x)
Traceback (most recent call last):
cannot compute leadterm(sqrt(1/x)*acosh(x + 1), x). The coefficient
should have been free of x but got sqrt(2)*sqrt(x)*sqrt(1/x)
```
Hence this addresses the issue along with giving us series at branch cuts` (-1, 1)`
```
>>> acosh(1 + x).series(x)
sqrt(2)*sqrt(x) - sqrt(2)*x**(3/2)/12 + 3*sqrt(2)*x**(5/2)/160 - 5*sqrt(2)*x**(7/2)/896 + 35*sqrt(2)*x**(9/2)/18432 - 63*sqrt(2)*x**(11/2)/90112 + O(x**6)
```
I don't see anything wrong or suspicious with the solution reported above ! Hence if you approve of this , I could implement this for one function atleast and `acosh` seems to be a good place to start , cause it has branch cuts along the real axis

Reply:
Does this also work with acosh(-1 + x)?

Reply:
Yes it does , I've checked related cases but just wanted some confidence behind the solution ! Thought of discussing it here as it's somewhat related !

Reply:
What is the result?

Reply:
```
>>> acosh(-1 + x).series(x)
I*pi - sqrt(2)*I*sqrt(x) - sqrt(2)*I*x**(3/2)/12 - 3*sqrt(2)*I*x**(5/2)/160 - 5*sqrt(2)*I*x**(7/2)/896 - 35*sqrt(2)*I*x**(9/2)/18432 - 63*sqrt(2)*I*x**(11/2)/90112 + O(x**6)
>>> 
>>> acosh(1 + x).series(x)
sqrt(2)*sqrt(x) - sqrt(2)*x**(3/2)/12 + 3*sqrt(2)*x**(5/2)/160 - 5*sqrt(2)*x**(7/2)/896 + 35*sqrt(2)*x**(9/2)/18432 - 63*sqrt(2)*x**(11/2)/90112 + O(x**6)
```
Result seem good when compared with wolfram !

Reply:
The expansion of acosh(-1 + x) has a branch cut for x < 0 where the function is single-valued.

Reply:
I agree that `acosh(x -1 )` has a branch cut from `(-oo, 0)` and `(0, 2)` but it's single valued after 2 right ? 
![image](https://user-images.githubusercontent.com/87052487/153692363-788c935b-f203-4777-998e-a244d3cffb82.png)
[Check out this wolfram link for reference](https://www.wolframalpha.com/input?i=limit+calculator&assumption=%7B%22F%22%2C+%22Limit%22%2C+%22limit%22%7D+-%3E%220%22&assumption=%7B%22F%22%2C+%22Limit%22%2C+%22limitfunction%22%7D+-%3E%22acosh%28x+-1%29%22) , cause we can see limit calculations supporting the point too .
We also have the correct result while calculating series at 2
```
>>> acosh(x - 1).series(x, 2)
sqrt(2)*sqrt(x - 2) - sqrt(2)*(x - 2)**(3/2)/12 + 3*sqrt(2)*(x - 2)**(5/2)/160 - 5*sqrt(2)*(x - 2)**(7/2)/896 + 35*sqrt(2)*(x - 2)**(9/2)/18432 - 63*sqrt(2)*(x - 2)**(11/2)/90112 + O((x - 2)**6, (x, 2))
```

Reply:
Also we know that `acosh(z) = ±i acos(z)`, where the sign depends on the location of `z` in the complex plane !
Hence the results seem quite relevant , don't you think so ?
```
>>> acos(x - 1).series(x)
pi - sqrt(2)*sqrt(x) - sqrt(2)*x**(3/2)/12 - 3*sqrt(2)*x**(5/2)/160 - 5*sqrt(2)*x**(7/2)/896 - 35*sqrt(2)*x**(9/2)/18432 - 63*sqrt(2)*x**(11/2)/90112 + O(x**6)
>>> acosh(x - 1).series(x)
I*pi - sqrt(2)*I*sqrt(x) - sqrt(2)*I*x**(3/2)/12 - 3*sqrt(2)*I*x**(5/2)/160 - 5*sqrt(2)*I*x**(7/2)/896 - 35*sqrt(2)*I*x**(9/2)/18432 - 63*sqrt(2)*I*x**(11/2)/90112 + O(x**6)
>>>
>>> 
>>> acos(x +1).series(x)
sqrt(2)*I*sqrt(x) - sqrt(2)*I*x**(3/2)/12 + 3*sqrt(2)*I*x**(5/2)/160 - 5*sqrt(2)*I*x**(7/2)/896 + 35*sqrt(2)*I*x**(9/2)/18432 - 63*sqrt(2)*I*x**(11/2)/90112 + O(x**6)
>>> acosh(x +1).series(x)
sqrt(2)*sqrt(x) - sqrt(2)*x**(3/2)/12 + 3*sqrt(2)*x**(5/2)/160 - 5*sqrt(2)*x**(7/2)/896 + 35*sqrt(2)*x**(9/2)/18432 - 63*sqrt(2)*x**(11/2)/90112 + O(x**6)
```

Reply:
Does this explanation convince you or am I missing something here ? I guess we could try this out !

Reply:
cc @jksuom